V5 Update Guide (#1601)
SAP/cloud-sdk: docs-java_versioned_docs/version-v5/guides/5.0-upgrade-steps.mdx

:::warning TODO

check for further packages and remove lines for deprecated modules

:::

| Package                                             | Module                                                                                   |
| --------------------------------------------------- | ---------------------------------------------------------------------------------------- |
| `com.sap.cloud.sdk.s4hana.serialization`            | `com.sap.cloud.sdk.s4hana:s4hana-core` (TODO: deprecated?)                               |
| `com.sap.cloud.sdk.s4hana.connectivity.exception`   | `com.sap.cloud.sdk.s4hana:s4hana-connectivity` (TODO: deprecated?)                       |
| `com.sap.cloud.sdk.s4hana.connectivity.rfc`         | `com.sap.cloud.sdk.s4hana:rfc` (TODO: deprecated?)                                       |
| `com.sap.cloud.sdk.s4hana.datamodel.odata.services` | See the section below on [generating your own OData VDM](#4-generate-your-own-odata-vdm) |
| `com.sap.cloud.sdk.cloudplatform.servlet`           | `com.sap.cloud.sdk.cloudplatform:servlet` dependency                                     |

> ## 4. Generate Your Own OData VDM

Reviewer: 3. Lacks the `s4hana-all` module
4. Could be more explicit by telling the user to add `odata-core` as a dependecy

Author: Yes, I wanted to have a look now which modules actually were removed/renamed, as the list on the other page seemed outdated.

---
id: 5.0-upgrade-steps
title: Steps to update to Version 5 of the SAP Cloud SDK
hide_title: false
hide_table_of_contents: false
sidebar_label: Update Steps to Version 5
description: Update your application to use version 5 of the SAP Cloud SDK
keywords:
  - sap
  - cloud
  - sdk
  - cloud native
  - cloud sdk
  - sap cloud sdk
  - 5
  - Update
---

This document will guide you, step-by-step on how you can leverage the new major version of the SAP Cloud SDK.

## 1. Change version of your sdk-bom or sdk-modules-bom dependency

The first step is to change the version of the SAP Cloud SDK dependency in your `pom.xml` file.
Depending on your project setup you might use one of our BOMs or the individual modules.

To find our BOMs, search for the following entries in your maven `dependencyManagement` section: `com.sap.cloud.sdk:sdk-bom` and `com.sap.cloud.sdk:sdk-modules-bom`.

> These you can update directly to version `5.0.0` or the [latest version](https://central.sonatype.com/artifact/com.sap.cloud.sdk/sdk-bom).

Reviewer 2: ```suggestion
Update them directly to version `5.0.0` or the [latest version](https://central.sonatype.com/artifact/com.sap.cloud.sdk/sdk-bom).
```

If you are using `com.sap.cloud.sdk:sdk-sjb-bom` you will need to replace it by using the contained BOMs directly:

```xml
<dependencyManagement>
    <dependencies>
      <!-- Replace usage of com.sap.cloud.sdk:sdk-sjb-bom by the following two BOMs -->
      <dependency>
        <groupId>com.sap.cloud.sjb.cf</groupId>
        <artifactId>cf-tomee7-bom</artifactId>
        <version>${your.sjb.version}></version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
      <dependency>
        <groupId>com.sap.cloud.sdk</groupId>
        <artifactId>sdk-bom</artifactId>
        <version>${sdk.v5.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
```

> ## 2. Adapt dependencies as needed

Reviewer 2: ```suggestion
## 2. Adapt Dependencies as Needed
```

> The lists below contains all modules that have been renamed or removed from the SDK without replacement.

Reviewer 2: [minor]

```suggestion
The lists below contains all modules that have been renamed or removed from the SDK.
```

  - `com.sap.cloud.sdk.cloudplatform:auditlog-scp-cf`
  - `com.sap.cloud.sdk.cloudplatform:auditlog-scp-neo`
  - `com.sap.cloud.sdk.frameworks`
    - child module `com.sap.cloud.sdk.frameworks:resilience4j` has been moved to `com.sap.cloud.sdk.cloudplatform:resilience4j`
    - child module `com.sap.cloud.sdk.frameworks:apache-httpclient5` has been moved to `com.sap.cloud.sdk.cloudplatform:connectivity-apache-httpclient5`
  - `com.sap.cloud.sdk.cloudplatform:cloudplatform-core-scp-cf`
  - The following modules' logic has been refactored into other modules, see release notes for details:
    - `com.sap.cloud.sdk.cloudplatform:security-dwc`
    - `com.sap.cloud.sdk.cloudplatform:security-scp-cf`
    - `com.sap.cloud.sdk.cloudplatform:tenant-dwc`
    - `com.sap.cloud.sdk.cloudplatform:tenant-scp-cf`

</details>

If you are using our BOMs to provide the module versions, a fast way to do identify those is to run any `mvn` command.
This will now lead to error messages like this:

> 'dependencies.dependency.version' for com.sap.cloud.sdk.cloudplatform:auditlog-scp-cf:jar is missing

This helps you to identify which modules you need to remove or replace in your project setup.

> ## 3. Introduce missing modules based on package name:

Reviewer 2: ```suggestion
## 3. Introduce Missing Modules Based on Package Name
```

:::warning TODO

check for further packages and remove lines for deprecated modules

:::

| Package                                             | Module                                                                                   |
| --------------------------------------------------- | ---------------------------------------------------------------------------------------- |
| `com.sap.cloud.sdk.s4hana.serialization`            | `com.sap.cloud.sdk.s4hana:s4hana-core` (TODO: deprecated?)                               |
| `com.sap.cloud.sdk.s4hana.connectivity.exception`   | `com.sap.cloud.sdk.s4hana:s4hana-connectivity` (TODO: deprecated?)                       |
| `com.sap.cloud.sdk.s4hana.connectivity.rfc`         | `com.sap.cloud.sdk.s4hana:rfc` (TODO: deprecated?)                                       |
| `com.sap.cloud.sdk.s4hana.datamodel.odata.services` | See the section below on [generating your own OData VDM](#4-generate-your-own-odata-vdm) |
| `com.sap.cloud.sdk.cloudplatform.servlet`           | `com.sap.cloud.sdk.cloudplatform:servlet` dependency                                     |

> ## 4. Generate your own OData VDM

Reviewer 2: ```suggestion
## 4. Generate Your own OData VDM
```

> To generate your own OData VDM, you can use the [OData Generator Maven Plugin](../features/odata/vdm-generator) directly.

Reviewer 2: I think I'd like to have a short paragraph (maybe just one sentence) about our decision to discontinue the pre-generated VDM and, therefore, customers have to switch to generating the VDM themselves.

You can use the following configuration to generate a VDM that is compatible with the OData VDM the SDK provided in version 4:

```xml
<inputDirectory>${project.basedir}/src/main/resources/odata/edmx</inputDirectory>
<outputDirectory>${project.build.directory}/vdm</outputDirectory>

> <serviceNameMappingFile>${project.basedir}/src/main/resources/odata/serviceNameMappings.properties</serviceNameMappingFile>

Reviewer 2: Should we (somehow?) provide our old `serviceNameMappings.properties` file? I think this would make it a lot easier for customers to migrate without breaking changes.

<deleteOutputDirectory>true</deleteOutputDirectory>
<packageName>com.sap.cloud.sdk.s4hana.datamodel.odata</packageName>
<compileScope>COMPILE</compileScope>
```

## 5. Handle OData VDM Instantiation

Depending on your project setup, you might have relied on the `@Inject` annotation in your OData VDM services.
As the Cloud SDK doesn't provide the pre-generated VDM anymore, you will need to instantiate your own services instead.
For that you can either use your own dependency injection framework, or instantiate the services manually.

The manual approach would look something like this:

```diff
- @Inject
- private BusinessPartnerService businessPartnerService;
+ private BusinessPartnerService businessPartnerService = new DefaultBusinessPartnerService();
```

> ## 6. Remove HttpFilter from your `web.xml`

Reviewer 2: ```suggestion
## 6. Remove HttpFilter from Your `web.xml`
```

> As we no longer provide the `HttpSecurityHeadersFilter` and `HttpCachingHeaderFilter` you can remove them from your `web.xml`.

Reviewer 2: Maybe add a brief explanation regarding that the XSUAA client lib should be used to protect applications.

---
id: 5.0-upgrade-steps
title: Steps to update to Version 5 of the SAP Cloud SDK
hide_title: false
hide_table_of_contents: false
sidebar_label: Update Steps to Version 5
description: Update your application to use version 5 of the SAP Cloud SDK
keywords:
  - sap
  - cloud
  - sdk
  - cloud native
  - cloud sdk
  - sap cloud sdk
  - 5
  - Update
---

This document provides a step-by-step guide on how you can leverage the new major version of the SAP Cloud SDK.

## 0. Before you Upgrade

> Before you increase the SAP Cloud SDK version to 5 it can be beneficial to migrate any usage of deprecated API first. Follow the deprecation note of the Javadoc in case you are still using deprecated API from the SAP Cloud SDK. All deprecated API is removed With version 5.

Reviewer 2: [minor] 

```suggestion
Before you increase the SAP Cloud SDK version to 5, it is recommended to first update to the latest available version of the SAP Cloud SDK 4 and migrate any usage of deprecated API. Follow the deprecation notes in the Javadocs to find recommended replacements. All deprecated APIs are removed with version 5.
```

> ## 1. Change Version of Your `sdk-bom` or `sdk-modules-bom` Bill of Materials

Reviewer 2: [minor]

_"bom"_ is the abbreviation for _"Bill of Materials"_ so the title is sort of dupliacted.

```suggestion
## 1. Change Version of Your `sdk-bom` or `sdk-modules-bom`
```

The first step is to change the version of the SAP Cloud SDK dependency in your `pom.xml` file.
Depending on your project setup you might use one of our BOMs or the individual modules.

To find our BOMs, search for the following entries in your maven `dependencyManagement` section: `com.sap.cloud.sdk:sdk-bom` and `com.sap.cloud.sdk:sdk-modules-bom`.
Update them directly to version `5.0.0` or the [latest version](https://central.sonatype.com/artifact/com.sap.cloud.sdk/sdk-bom).

> If you are using `com.sap.cloud.sdk:sdk-sjb-bom` you will either need to replace it with one of the BOMs mentioned above, or stay on the latest version 4 of the SAP Cloud SDK.

Reviewer 2: This sentence still sounds like all SJB customers could update the SDK, which is, unfortunately, not true for customers that rely on JCo (or `war` deployment, more generally).

## 2. Adapt Dependencies as Needed

The lists below contains all modules that have been renamed or removed from the SDK.
Therefore, you should also rename respectively remove those from your project setup.

<details>
  <summary>Full list of renamed modules</summary>

- Module `com.sap.cloud.sdk.frameworks:resilience4j` has been moved to `com.sap.cloud.sdk.cloudplatform:resilience4j`
- Module `com.sap.cloud.sdk.frameworks:apache-httpclient5` has been moved to `com.sap.cloud.sdk.cloudplatform:connectivity-apache-httpclient5`

</details>

<details>
    <summary>Full list of removed modules</summary>

- Bill of Materials

>   - `com.sap.cloud.sdk:sdk-sjb-bom

Reviewer 2: ```suggestion
  - `com.sap.cloud.sdk:sdk-sjb-bom`
```

  - `com.sap.cloud.sdk.datamodel:openapi-generator-cli`
- Pregenerated Clients
  - `com.sap.cloud.sdk.s4hana:s4hana-all`
  - `com.sap.cloud.sdk.s4hana:s4hana-api-odata`
  - `com.sap.cloud.sdk.s4hana:s4hana-api-odata-onpremise`
  - `com.sap.cloud.sdk.s4hana:s4hana-api-odata-v4`
  - `com.sap.cloud.sdk.s4hana:s4hana-api-odata-v4-onpremise`
- Services
  - `com.sap.cloud.sdk.services:blockchain-business-services-sdk`
  - `com.sap.cloud.sdk.services:blockchain-business-services-visibility`
  - `com.sap.cloud.sdk.services:blockchain-client-fabric`
  - `com.sap.cloud.sdk.services:blockchain-client-multichain`
  - `com.sap.cloud.sdk.services:btp-business-rules`
  - `com.sap.cloud.sdk.services:btp-shared-ledger-client`
  - `com.sap.cloud.sdk.services:mdi-client-convenience`
  - `com.sap.cloud.sdk.services:mdi-client-core`
  - `com.sap.cloud.sdk.services:mdi-client-model`
  - `com.sap.cloud.sdk.services:mdi-client-parent`
  - `com.sap.cloud.sdk.services:mdi-generator`
  - `com.sap.cloud.sdk.services:mdi-model-core`
  - `com.sap.cloud.sdk.services:mdo-distribution-client-api`
  - `com.sap.cloud.sdk.services:mdo-distribution-client-integration-test`

>   - `com.sap.cloud.sdk.services:mdo-distribution-client-odata`

Reviewer 2: The MDI/MDO modules are internal only. Therefore, we never mentioned them in our Documentation portal... Not sure whether we should now state that they have been removed over here? 
Personally I'd say: No, we have already announced the deprecation and removal in the MDI docs.

Author: I think this question also applies for the complete "SAP Internal" section, right? I moved MDI/O over to that section to keep the discussion together.

:::warning TODO

check for further packages and remove lines for deprecated modules

:::

| Package                                             | Module                                                                                   |
| --------------------------------------------------- | ---------------------------------------------------------------------------------------- |
| `com.sap.cloud.sdk.s4hana.serialization`            | `com.sap.cloud.sdk.s4hana:s4hana-core` (TODO: deprecated?)                               |
| `com.sap.cloud.sdk.s4hana.connectivity.exception`   | `com.sap.cloud.sdk.s4hana:s4hana-connectivity` (TODO: deprecated?)                       |
| `com.sap.cloud.sdk.s4hana.connectivity.rfc`         | `com.sap.cloud.sdk.s4hana:rfc` (TODO: deprecated?)                                       |
| `com.sap.cloud.sdk.s4hana.datamodel.odata.services` | See the section below on [generating your own OData VDM](#4-generate-your-own-odata-vdm) |
| `com.sap.cloud.sdk.cloudplatform.servlet`           | `com.sap.cloud.sdk.cloudplatform:servlet` dependency                                     |

## 4. Generate Your Own OData VDM

> The SAP Cloud SDK no longer provides a pre-generated OData VDM. We made this decision to reduce the maintenance effort and to allow you to generate your own VDM with the exact version of the OData service you are using. That way there is no need to wait for the next SDK release to get the latest changes of the OData service.

Reviewer 2: [minor]

```suggestion
The SAP Cloud SDK no longer provides pre-generated OData VDMs. We made this decision to reduce the maintenance effort and to allow you to generate your own VDM with the exact version of the OData service you are using. That way there is no need to wait for the next SDK release to get the latest changes of an OData service you are consuming.
```

---
id: 5.0-upgrade-steps
title: Steps to update to Version 5 of the SAP Cloud SDK
hide_title: false
hide_table_of_contents: false
sidebar_label: Update Steps to Version 5
description: Update your application to use version 5 of the SAP Cloud SDK
keywords:
  - sap
  - cloud
  - sdk
  - cloud native
  - cloud sdk
  - sap cloud sdk
  - 5
  - Update
---

This document provides a step-by-step guide on how you can leverage the new major version of the SAP Cloud SDK.

## 0. Before you Upgrade

Before you increase the SAP Cloud SDK version to 5, it is recommended to first update to the latest available version of the SAP Cloud SDK 4 and migrate any usage of deprecated API. Follow the deprecation notes in the Javadocs to find recommended replacements. All deprecated APIs are removed with version 5.

Also, make sure that you are using Java 17 or higher. Besides changing the actual compilation binary you should also adapt the following maven properties to `17` in your `pom.xml` file:

> - `java.version`

Reviewer 3: I think this is not entirely correct. I would just link to some docu or guide from Maven on how to do this, e.g. [here](https://www.baeldung.com/maven-java-version).

Reviewer 3: Also, it might be good to highlight only Java 17 is the prerequisite (and not just a recommendation) but the Spring Stack can be updated in the next step (maybe this requires including the javax.servlet dependency temporarily..). Otherwise users may struggle to upgrade to the Spring 6 stack while remaining on SDK 4...

Reviewer 3: Maybe we should mention the minimum required CAP version for CAP users here too?

Author: Makes complete sense. I'm a bit out of the CAP loop: What is the version we support/require with our version 5? I don't see direct dependencies on our side, so I guess it's related to changes we did some weeks ago regarding the v3/v4 check?

Reviewer 3: If you are using CAP, [this statement](https://github.com/SAP/cloud-sdk/blob/main/docs-java_versioned_docs/version-v5/guides/5.0-upgrade.mdx#L28) applies.

  - `com.sap.cloud.sdk.cloudplatform:security-servlet`
  - `com.sap.cloud.sdk.cloudplatform:tenant-scp-cf`

</details>

If you are using our BOMs to provide the module versions, a fast way to do identify those is to run any `mvn` command.
This will now lead to error messages like this:

> 'dependencies.dependency.version' for com.sap.cloud.sdk.cloudplatform:auditlog-scp-cf:jar is missing

This helps you to identify which modules you need to remove or replace in your project setup.

## 3. Introduce Missing Modules Based on Package Names

Once the build doesn't complain about missing dependency versions anymore, you can start to fix the compilation errors.
For this, the following table provides a helpful overview, over which packages can be found in which module.

:::warning TODO

check for further packages and remove lines for deprecated modules

:::

| Package                                             | Module                                                                                   |
| --------------------------------------------------- | ---------------------------------------------------------------------------------------- |

> | `com.sap.cloud.sdk.s4hana.serialization`            | `com.sap.cloud.sdk.s4hana:s4hana-core` (TODO: deprecated?)                               |

Reviewer 3: This list is somewhat unclear to me, is this representing the old v4 or the new v5 state? Eitherway, what should I do with this information as a reader?

From the title I would have expected something like `Package Name | Old Maven Module | New Maven Module`. Assuming this section is only about stuff that changed modules but not packages

            <artifactId>sdk-bom</artifactId>
            <version>5.0.0</version> <!-- TODO: replace this with the most recent version -->
            <type>pom</type>
            <scope>import</scope>
        </dependency>
    </dependencies>
</dependencyManagement>
```

At the same time you should also replace usages of `javax` dependencies with the respective `jakarta` dependency. For example:

```diff
-<dependency>
-    <groupId>javax.inject</groupId>
-    <artifactId>javax.inject</artifactId>
-</dependency>
+<dependency>
+    <groupId>jakarta.inject</groupId>
+    <artifactId>jakarta.inject-api</artifactId>
+</dependency>
```

## Remove or Replace Modified Cloud SDK Dependencies

The lists below contains all modules that have been renamed or removed from the SDK.
Therefore, you should also rename respectively remove those from your project setup.

<details>
  <summary>Full list of renamed or replaced modules</summary>

- Module `com.sap.cloud.sdk.frameworks:resilience4j` has been moved to `com.sap.cloud.sdk.cloudplatform:resilience4j`
- Module `com.sap.cloud.sdk.frameworks:apache-httpclient5` has been moved to `com.sap.cloud.sdk.cloudplatform:connectivity-apache-httpclient5`

> - Module `com.sap.cloud.sdk.cloudplatform:servlet` has been removed and functionally replaced by `com.sap.cloud.sdk.cloudplatform:servlet-jakarta` (replacing usage of `javax.servlet` with `jakarta.servlet`)

Reviewer: ```suggestion
- Module `com.sap.cloud.sdk.cloudplatform:servlet` has been removed and functionally replaced by `com.sap.cloud.sdk.cloudplatform:servlet-jakarta` (replacing usage of `javax.servlet` with `jakarta.servlet`)
- Module `com.sap.cloud.sdk.cloudplatform:scp-cf` has been removed and functionally replaced by `com.sap.cloud.sdk:sdk-core`
- Module `com.sap.cloud.sdk.cloudplatform:dwc-cf` has been removed and functionally replaced by both `com.sap.cloud.sdk:sdk-core` and `com.sap.cloud.sdk.cloudplatform:connectivity-dwc`
```

Author: I would keep the dwc related topics out of this guide here, as discussed earlier this week in standup. The other point makes sense.

</details>

<details>
    <summary>Full list of removed modules</summary>

- Bill of Materials

>   - `com.sap.cloud.sdk:sdk-sjb-bom`

Reviewer: This list should have a reason behind every removal:
A replacement module or explanation on where to go from now on

For this, the following table provides a helpful overview, over which packages can be found in which module.

:::warning TODO

check for further packages and remove lines for deprecated modules

:::

| Package                                             | Module                                                                                   |
| --------------------------------------------------- | ---------------------------------------------------------------------------------------- |
| `com.sap.cloud.sdk.s4hana.serialization`            | `com.sap.cloud.sdk.datamodel:odata-core`                                                 |
| `com.sap.cloud.sdk.s4hana.connectivity.exception`   | `com.sap.cloud.sdk.s4hana:s4hana-connectivity` (TODO: deprecated?)                       |
| `com.sap.cloud.sdk.s4hana.connectivity.rfc`         | `com.sap.cloud.sdk.s4hana:rfc` (TODO: deprecated?)                                       |
| `com.sap.cloud.sdk.s4hana.datamodel.odata.services` | See the section below on [generating your own OData VDM](#4-generate-your-own-odata-vdm) |

> | `com.sap.cloud.sdk.cloudplatform.servlet`           | `com.sap.cloud.sdk.cloudplatform:servlet` dependency                                     |

Reviewer: The servlet module has been removed and is replaced with servlet-jakarta, which is now stable

### Handle OData VDM Instantiation

Depending on your project setup, you might have relied on the `@Inject` annotation in your OData VDM services.
As the OData generators don't generate the `javax.inject.Named` annotations anymore, you will need to instantiate your own services instead.
For that you can either use your own dependency injection framework, or instantiate the services manually.

The manual approach would look something like this:

```diff
- @Inject
- private BusinessPartnerService businessPartnerService;
+ private BusinessPartnerService businessPartnerService = new DefaultBusinessPartnerService();
```

## Remove Discontinued ServletFilter from Your `web.xml`

We decided to remove all security related `Filter` implementations from the SAP Cloud SDK, to have a clearer separation between the [SAP Security Libraries](https://github.com/SAP/cloud-security-services-integration-library/) and the SAP Cloud SDK.

This means you need to remove all mentions of the following filters from your `web.xml` (mind the `security` in the package): `com.sap.cloud.sdk.cloudplatform.security.servlet.*Filter`

> The `com.sap.cloud.sdk.cloudplatform.servlet.RequestAccessorFilter` remains in the SDK, as it is not security related.

Reviewer: Removed from the SDK, replaced by `com.sap.cloud.sdk.cloudplatform.servletjakarta.RequestAccessorFilter`

  - `com.sap.cloud.sdk.services:scp-blockchain`
  - `com.sap.cloud.sdk.services:scp-business-logging-all`
  - `com.sap.cloud.sdk.services:scp-business-logging-core`
  - `com.sap.cloud.sdk.services:scp-business-logging-odata`
  - `com.sap.cloud.sdk.services:scp-business-logging-rest`
  - `com.sap.cloud.sdk.services:scp-workflow-cf`
- Others
  - `com.sap.cloud.sdk.cloudplatform:auditlog` (See `AuditLogger` in the <a href="#replaceremove-changed-api">list of deprecated API</a> section below)
  - `com.sap.cloud.sdk.cloudplatform:auditlog-scp-cf` (See `AuditLogger` in the <a href="#replaceremove-changed-api">list of deprecated API</a> section below)
  - `com.sap.cloud.sdk.cloudplatform:cloudplatform-connectivity-scp` (No replacement necessary)
  - `com.sap.cloud.sdk.cloudplatform:cloudplatform-core-scp-cf` ()
  - `com.sap.cloud.sdk.cloudplatform:security-scp-cf` (Functionally integrated into the `com.sap.cloud.sdk.cloudplatform:security` module)
  - `com.sap.cloud.sdk.cloudplatform:security-servlet` (See the <a href="">ServletFilter</a> section below)
  - `com.sap.cloud.sdk.cloudplatform:tenant-scp-cf` (Functionally integrated into the `com.sap.cloud.sdk.cloudplatform:tenant` module)

>     - Replaced by `com.sap.cloud.sdk.cloudplatform:tenant`

Reviewer: ```suggestion
```

Author: Ups, that must have slipped through the merge. Thanks! Will change it locally.

:::

:::info Minimum CAP version
If you are using the Cloud Application Programming model (CAP) to build your application, please note that the SAP Cloud SDK version 5 requires a minimum CAP/CDS version of at least `2.4.0`.
If you are on an older CAP stack, please follow the [CAP migration guide](https://cap.cloud.sap/docs/java/migration#one-to-two) first.
:::

## Before You Update

Before you increase the SAP Cloud SDK version to 5, it is recommended to first update to the latest available version of the SAP Cloud SDK 4 and migrate any usage of deprecated API. Follow the deprecation notes in the Javadocs to find recommended replacements. All deprecated APIs are removed with version 5.

Also, make sure that you are using Java 17 or higher. To setup your build to use Java 17 you can use [this guide](https://www.baeldung.com/maven-java-version).
As Java EE modules have been removed from Java 11, you might also need to update your application to use the [Jakarta EE](https://jakarta.ee/) modules instead.
This JEP provides a good overview of the changes: [JEP 320: Remove the Java EE and CORBA Modules](https://openjdk.java.net/jeps/320).

> 

Reviewer 4: ```suggestion
If you are consuming SOAP APIs and are relying on the `axis2-wsdl2code-maven-plugin` to generate classes, please also additionally include `javax.activation:activation` artifact to your dependencies to avoid compilation errors.
```

:::warning TODO

When to update java and spring? Together or separately?

> :::

Reviewer 4: I updated java first and then Spring and the Cloud SDK together. 
After updating java because of the inclusion generated classes for soap module, I got a compilation error: `cannot access javax.activation.DataHandler` for which I had to include the `javax.activation:activation` artifact.

These BOMs should be listed after any SAP Cloud SDK related BOM, so for Spring Boot your `dependencyManagement` section should look something like this:

```xml
<dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>com.sap.cloud.sdk</groupId>
            <artifactId>sdk-bom</artifactId>
            <version>5.0.0</version> <!-- TODO: replace this with the most recent version -->
            <type>pom</type>
            <scope>import</scope>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-dependencies</artifactId>
            <version>3.1.6</version> <!-- TODO: replace this with the most recent version -->
            <type>pom</type>
            <scope>import</scope>
        </dependency>
    </dependencies>
</dependencyManagement>
```

At the same time you should also replace usages of `javax` dependencies with the respective `jakarta` dependency:

```diff
-<dependency>
-    <groupId>javax.inject</groupId>
-    <artifactId>javax.inject</artifactId>
-</dependency>
+<dependency>
+    <groupId>jakarta.inject</groupId>
+    <artifactId>jakarta.inject-api</artifactId>

> +</dependency>

Reviewer 4: We don't manage the `jakarta.inject:jakarta.inject-api` in our BOM, although we used to manage `javax.inject:javax.inject`. So, the users have to explicitly add in a version.

> ```

Author: Hm, but why was that necessary? Is the app running on Java8? There should be a `jakarta.inject.Named` annotation available, right? Or is that just to be able to run with SDK v4?

> ```

Reviewer 3: That is an RFC class that was copied over.. Probably we should have just removed the `@Named` from that code. It's not generated IIRC, we only have a generator for SOAP.

> ```

Reviewer: Then I will remove it from the dwc-test-app

> ```

Reviewer: But customers need to be aware that we removed the dependency from our bom.

> ```

Reviewer: Please review [Removed javax.inject dependency](https://github.tools.sap/cloudsdk/dwc-cap-orbits/pull/20#top)

  - `com.sap.cloud.sdk.cloudplatform:auditlog-scp-neo`
  - `com.sap.cloud.sdk.cloudplatform:cloudplatform-connectivity-scp-neo`
  - `com.sap.cloud.sdk.cloudplatform:cloudplatform-core-scp-neo`
  - `com.sap.cloud.sdk.cloudplatform:concurrency-scp-neo`
  - `com.sap.cloud.sdk.cloudplatform:scp-neo`
  - `com.sap.cloud.sdk.cloudplatform:security-scp-neo`
  - `com.sap.cloud.sdk.cloudplatform:tenant-scp-neo`
  - `com.sap.cloud.sdk.plugins:scp-neo-maven-plugin`
- Generator CLIs
  - `com.sap.cloud.sdk.datamodel:odata-generator-cli`
    - Call the generator directly as a maven plugin:
    - `mvn com.sap.cloud.sdk.datamodel:odata-generator-maven-plugin:5.0.0:generate`
  - `com.sap.cloud.sdk.datamodel:odata-v4-generator-cli`
    - Call the generator directly as a maven plugin:
    - `mvn com.sap.cloud.sdk.datamodel:odata-v4-generator-maven-plugin:5.0.0:generate`
  - `com.sap.cloud.sdk.datamodel:openapi-generator-cli`
    - Call the generator directly as a maven plugin:
    - `mvn com.sap.cloud.sdk.datamodel:openapi-generator-maven-plugin:5.0.0:generate`
- Pregenerated Clients. [Generate and use your own OData VDM](#generate-and-use-your-own-odata-vdm) instead.

>   - `com.sap.cloud.sdk.s4hana:s4hana-all`

Reviewer 4: As the Spring E2E test also tests RFC calls, when `com.sap.cloud.sdk.s4hana:s4hana-all` was removed, I also had to additionally add `com.sap.cloud.sdk.s4hana:s4hana-connectivity` and `com.sap.cloud.sdk.s4hana:rfc` for the project to compile.

Author: As far as I know RFC is no longer supported, so that the related changes are not part of the update guide.

Reviewer 4: True, my bad, it is no longer released with v5.

        <td>No replacement.</td>
      </tr>
      <tr>
        <td>ThrowableAssertionUtil</td>
        <td>-</td>
        <td>No replacement.</td>
      </tr>
    </tbody>
  </table>
</details>

## Generate and Use Your Own OData VDM

### Generate Your Own OData VDM

The SAP Cloud SDK no longer provides pre-generated OData VDMs. We made this decision to reduce the maintenance effort and to allow you to generate your own VDM with the exact version of the OData service you are using. That way there is no need to wait for the next SDK release to get the latest changes of an OData service you are consuming.

To generate your own OData VDM, you can use the [OData Generator Maven Plugin](../features/odata/vdm-generator) directly.
You can use the following configuration to generate a VDM that is compatible with the OData VDM the SDK provided in version 4:

```xml
<inputDirectory>${project.basedir}/src/main/resources/odata/edmx</inputDirectory>
<outputDirectory>${project.build.directory}/vdm</outputDirectory>
<serviceNameMappingFile>${project.basedir}/src/main/resources/odata/serviceNameMappings.properties</serviceNameMappingFile>
<deleteOutputDirectory>true</deleteOutputDirectory>

> <packageName>com.sap.cloud.sdk.s4hana.datamodel.odata</packageName>

Reviewer 4: Package name is different for OData v4. Maybe having a tab for `v4` is easier.
Additionally, package names are also different for `onpremise` too. Adding a note to that the users using on-premise services have to adjust their imports would help.

<compileScope>COMPILE</compileScope>
```

The file specified under _serviceNameMappingFile_ will be created and populated, if it doesn't exist yet.

> If you want to create a VDM that is compatible with the OData VDM the SDK provided in version 4, you can use this [_serviceNameMappings.properties_](../../../static/resources/v4-serviceNameMappings.properties).

Reviewer 4: Doesn't contain the on-premise entries or OData v4 entries.
When I downloaded the specification from the Accelerator hub, the Business Partner spec was named(by default) as `OP_API_BUSINESS_PARTNER.edmx`. So, I had to rename it, to generate classes without any renaming. 
Which raises the question for other Onpremise/OData v4 services that may not have an entry in this file.

If you are on an older CAP stack, please follow the [CAP migration guide](https://cap.cloud.sap/docs/java/migration#one-to-two) first.
:::

## Before You Update

Before you increase the SAP Cloud SDK version to 5, it is recommended to first update to the latest available version of the SAP Cloud SDK 4 and migrate any usage of deprecated API. Follow the deprecation notes in the Javadocs to find recommended replacements. All deprecated APIs are removed with version 5.

Also, make sure that you are using Java 17 or higher. To setup your build to use Java 17 you can use [this guide](https://www.baeldung.com/maven-java-version).
As Java EE modules have been removed from Java 11, you might also need to update your application to use the [Jakarta EE](https://jakarta.ee/) modules instead.
This JEP provides a good overview of the changes: [JEP 320: Remove the Java EE and CORBA Modules](https://openjdk.java.net/jeps/320).

If you are consuming SOAP APIs and are relying on the `axis2-wsdl2code-maven-plugin` to generate classes, please also additionally include `javax.activation:activation` artifact to your dependencies to avoid compilation errors.

> 

Reviewer 4: The order of update note, first Java followed by SDK+Spring together seems to have been dropped somehow.
I think it would still be beneficial to have this as a `tip`.

Author: Fair, it's implicitly mentioned by the order of the sections, but a concrete hint will be helpful.

      </tr>
      <tr>
        <td>StringCollectionPrincipalAttribute</td>
        <td><a href= "https://github.com/SAP/cloud-security-services-integration-library/tree/main">
                            SAP BTP Security Services Integration Libraries</a></td>
        <td>See comment on <a href="#XsuaaComment">retrieving token information.</a></td>
      </tr>
      <tr>
        <td>StringPrincipalAttribute</td>
        <td><a href= "https://github.com/SAP/cloud-security-services-integration-library/tree/main">
                            SAP BTP Security Services Integration Libraries</a></td>
        <td>See comment on <a href="#XsuaaComment">retrieving token information.</a></td>
      </tr>
      <tr>
        <td>ScpCfDestinationRetrievalStrategy.CURRENT_TENANT_THEN_PROVIDER</td>
        <td>Please query subscriber and provider tenants individually instead using <code>ONLY_SUBSCRIBER</code> and <code>ALWAYS_PROVIDER</code></td>
        <td>-</td>
      </tr>
      <tr>
        <td>PrincipalPropagationStrategy.setDefaultStrategy(...)</td>
        <td>Destination custom property: <br /><code>cloudsdk.principalPropagationMode = (recommended|compatibility)</code></td>

>         <td>Default: <code>compatibility</code>, for reduced token exchanges. <a href="https://help.sap.com/docs/connectivity/sap-btp-connectivity-cf/configure-principal-propagation-via-user-exchange-token" target="_new">See more.</a></td>

Reviewer 3: ```suggestion
        <td>Default: <code>TOKEN_FORWARDING</code>, for reduced token exchanges. <a href="https://help.sap.com/docs/connectivity/sap-btp-connectivity-cf/configure-principal-propagation-via-user-exchange-token" target="_new">See more.</a></td>
```
